File: python/construct_string_from_binary_tree.py
# no. 606
class Solution:
    def __init__(self):
        self.ans = ''
    def preorder_traverse(self, root):
      if root:
        self.ans += f'{root.val}'
      if root.left:
        self.ans += '('
        self.preorder_traverse(root.left)
        self.ans += ')'
      if root.right:
        if not root.left:
          self.ans += '()'
        self.ans += '('
        self.preorder_traverse(root.right)
        self.ans += ')'
      if not root.left and not root.right:
        return

    def tree2str(self, root: Optional[TreeNode]) -> str:
        self.preorder_traverse(root)
        return self.ans

File: C++/construct_string_from_binary_tree.cpp
//no. 606
#include "tree_node.h"
#include <string>
using namespace std;

class Solution {
public:
    string ans;
    void preorder_traverse(TreeNode* root){
      if(root){
        ans += to_string(root->val);
      }
      if(root->left){
        ans += "(";
        preorder_traverse(root->left);
        ans += ")";
      }
      if(root->right){
        if(!root->left){
          ans += "()";
        }
        ans += "(";
        preorder_traverse(root->right);
        ans += ")";
      }
      if(!root->right & !root->left){
        return;
      }
    }
    string tree2str(TreeNode* root) {
      preorder_traverse(root);
      return ans;
    }
};
